ci: coccinelle: Add checker for rules 5.7, 14.4 and 21.2

ceolin opened the pull request:
This can spot most violations recently fixed in 
https://github.com/zephyrproject-rtos/zephyr/pull/33569
https://github.com/zephyrproject-rtos/zephyr/pull/33562

Commit:
scripts: coccinelle: Check rule 5.7

This coccinelle script can check some violations for rule 5.7.
It can identify things like:

struct device *device

But it is not capable to identify:

struct test { ... }

...

int test;

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
scripts: coccinelle: Check rule 21.2

This coccinelle script can check some violations for rule 21.2.
Currently it is checking the follow reserved names:

"remove", "rewind", "malloc", "free", "exp", "signal"

It can easily be extended in the future though.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Rebnase

Commit:
ci: Apply coccinelle scripts in git diffs

This scripts receives the same parameter of what_changed.py. And run
coccinelle scripts for code guideline compliance in the given git
commits. e.g: ./guideline_check.py --commits origin/master..HEAD

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

nashif commented on scripts/ci/guideline_check.py:
```diff
@@ -0,0 +1,82 @@
+def main():
+    args = parse_args()
+    if not args.commits:
+        exit(1)
```
exit("missing commit range")

nashif commented on scripts/ci/guideline_check.py:
```diff
@@ -0,0 +1,83 @@
+
+
+if __name__ == "__main__":
+    main()
```
main() should return something else other 0 if the scan fails and violations are found.

nashif commented on scripts/ci/guideline_check.py:
```diff
@@ -0,0 +1,83 @@
+                if line.is_added:
+                    violation = "{}:{}".format(f.path, line.target_line_no)
+                    if violation in violations:
+                        print(
```
we should add an option to send the output to a file.

nashif reviewed (changes requested):
see https://github.com/nashif/zephyr/commit/5165f48cca4318f348cccd9307c6d1df2bb541e0 and  https://github.com/nashif/zephyr/commit/1ae7e44963bc3efce0d97dd5d883ce4f807ce351

Commit:
scripts: coccinelle: Add guideline number on a script output

Add information about the guideline number in the script checking a
identifier length.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
ci: coccinelle: Add support to write to a file

Add parameter to write the script output to a file.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
ci: coccinelle: Suppress pylint false positive

Suppress pylint false positive on 'sh' library.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

nashif commented on scripts/ci/guideline_check.py:
```diff
@@ -0,0 +1,98 @@
+                                    violation, "\t\n".join(
+                                        violations[violation])))
+
+        return numViolations
```
indent here is wrong

ceolin replied:
arghh, good catch

Commit:
coccinelle: Add script to check rule 14.4

Controlling expression shall have essentially boolean
type.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
ci: guideline: Run script to check rule 14.4

Run script to check violations to rule 14.4

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

jenmwms commented on scripts/ci/guideline_check.py:
```diff
@@ -0,0 +1,99 @@
+        exit("missing commit range")
+
+    # pylint does not like the 'sh' library
+    # pylint: disable=too-many-function-args,unexpected-keyword-arg
```
Is this command for pylint disable left over from debug perhaps? Or is this a suggestion in special scenarios?

ceolin replied:
nope, this is due a pylint issue.

jenmwms replied:
I see the commit adding this says "suppress pylint flase positive on 'sh' library". Does this commented line execute?
